StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/FixAllContextHelper.cs

```diff
@@ -0,0 +1,228 @@
+                        }
+                    });
+
+                    return projectsAndDiagnostics.ToImmutableDictionary();
+                }
+            }
+
+            return ImmutableDictionary<Project, ImmutableArray<Diagnostic>>.Empty;
+        }
+
+        private static ImmutableDictionary<string, ImmutableArray<Type>> GetAllAnalyzers()
+        {
+            Assembly assembly = typeof(NoCodeFixAttribute).GetTypeInfo().Assembly;
+
+            var diagnosticAnalyzerType = typeof(DiagnosticAnalyzer);
+
+            var analyzers = ImmutableDictionary.CreateBuilder<string, ImmutableArray<Type>>();
+
+            foreach (var type in assembly.DefinedTypes)
+            {
+                if (type.IsSubclassOf(diagnosticAnalyzerType) && !type.IsAbstract)
+                {
+                    Type analyzerType = type.AsType();
+                    DiagnosticAnalyzer analyzer = (DiagnosticAnalyzer)Activator.CreateInstance(analyzerType);
+                    foreach (var descriptor in analyzer.SupportedDiagnostics)
+                    {
+                        ImmutableArray<Type> types;
+                        if (analyzers.TryGetValue(descriptor.Id, out types))
+                        {
+                            types = types.Add(analyzerType);
+                        }
+                        else
+                        {
+                            types = ImmutableArray.Create(analyzerType);
+                        }
+
+                        analyzers[descriptor.Id] = types;
+                    }
+                }
+            }
+
+            return analyzers.ToImmutable();
+        }
+
+        private static ImmutableArray<DiagnosticAnalyzer> GetDiagnosticAnalyzersForContext(FixAllContext fixAllContext)
+        {
+            return DiagnosticAnalyzers
+                .Where(x => fixAllContext.DiagnosticIds.Contains(x.Key))
+                .SelectMany(x => x.Value)
+                .Distinct()
+                .Select(type => (DiagnosticAnalyzer)Activator.CreateInstance(type))
+                .ToImmutableArray();
+        }
+
+        private static async Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync(FixAllContext fixAllContext, Project project, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            bool includeCompilerDiagnostics = fixAllContext.DiagnosticIds.Any(x => x.StartsWith("CS", StringComparison.Ordinal));
+            var compilation = await project.GetCompilationAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+
+            var diagnostics = ImmutableArray<Diagnostic>.Empty;
+            if (analyzers.Any())
+            {
+                var compilationWithAnalyzers = compilation.WithAnalyzers(analyzers, project.AnalyzerOptions, fixAllContext.CancellationToken);
+                diagnostics = await compilationWithAnalyzers.GetAnalyzerDiagnosticsAsync().ConfigureAwait(false);
```

Comment:
GetAnalyzerDiagnostics seems to be affected by the same bug you pointed out in your churn testing. You may try to replicate what IDE diagnostic service does sequentially - for each document invoke GetAnalyzerSyntaxDiagnosticsAsync and GetAnalyzerSemanticDiagnosticsAsync. If you also have any compilation end analyzers, then also invoke GetAnalyzerCompilationDiagnosticsAsync at the end. This won't be as fast as GetAnalyzerDiagnosticsAsync() but should be faster then current performance of FixAllContext.GetAllDiagnosticsAsync.